### PR TITLE
Workaround for platforms lacking crt_destructor

### PR DESCRIPTION
**Pull request by maxhaton**
This will eventually fix https://issues.dlang.org/show_bug.cgi?id=22520 however not tagging just yet as it doesn't specifically target that platform yet.

The mitigation is currently set to always on so I can run it through MacOS CI (so windows will presumably fail).

### CODE REVIEW

File: src/dmd/glue.d [hunk 1]
```diff
@@ -1163,8 +1163,67 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
+            //Global SC for now.
+            newConstructor.Sclass = SCglobal;
+            func_t* funcState = newConstructor.Sfunc;
+            //Init start block
+            funcState.Fstartblock = block_calloc();
+            block* startBlk = funcState.Fstartblock;
+            //Make that block run __cxa_atexit(&func);
+            auto atexitSym = getRtlsym(RTLSYM.CXA_ATEXIT);
+            //Build parameter pack - __cxa_atexit(&func, null, null)
+            auto paramPack = el_params(el_long(TYnptr, 0), el_long(TYnptr, 0), el_ptr(s), null);
```
Comment (ibuclaw) [inline src/dmd/glue.d:1210]: To make it friendly for shared libraries:
```
__cxa_atexit(&func, null, &__dso_handle);
```
`__dso_handle` itself is a weak symbol defined in every at least every dso object that cares about Itanium/C++ compatibility.

File: src/dmd/glue.d [hunk 2]
```diff
@@ -1163,8 +1163,67 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
     if (fd.isCrtCtorDtor & 2)
-        objmod.setModuleCtorDtor(s, false);
+    {
+        bool mitigation = true;
+        /*
+            https://issues.dlang.org/show_bug.cgi?id=22520
+
+            Apple radar: https://openradar.appspot.com/FB9733712
+
+            Apple deprecated the mechanism used to implement `crt_constructor`
```
Comment (ibuclaw) [inline src/dmd/glue.d:1175]:
```suggestion
            Apple deprecated the mechanism used to implement `crt_destructor`
```

File: src/dmd/glue.d [hunk 3]
```diff
@@ -1163,8 +1163,67 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
 
     if (fd.isCrtCtorDtor & 1)
         objmod.setModuleCtorDtor(s, true);
+
     if (fd.isCrtCtorDtor & 2)
-        objmod.setModuleCtorDtor(s, false);
+    {
+        bool mitigation = true;
```
Comment (ibuclaw) [inline src/dmd/glue.d:1169]: TargetC is the best place for this, something like:
```
if (target.c.dtorsUseAtexit)
```

**Comment by ibuclaw** [comment]
> The mitigation is currently set to always on so I can run it through MacOS CI (so windows will presumably fail).

You could add the Monterey pipeline to Cirrus to validate this too.

File: src/dmd/glue.d (hunk 2 above)
Comment (maxhaton) [reply src/dmd/glue.d:1175]: Oops!

**Review by ljmf00 — DISMISSED** [review]
Missing tests

**Comment by maxhaton** [comment]
Why? There are tests for crt_... already, this doesn't add anything

**Comment by ibuclaw** [comment]
> Missing tests

Tests are failing, which is why this bug report was raised.

File: src/dmd/glue.d [hunk 4]
```diff
@@ -1163,8 +1163,67 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
+            the destructor-to-be using __cxa_atexit()
+
+            This workaround may need a further look at when it comes to
+            shared library support, however there is no bridge for
+            that spilt milk to flow under yet.
+
+            This relies on the Itanium ABI so is portable to any
+            platform it, if needed.
+        */
+        if(mitigation)
```
Comment (ljmf00) [inline src/dmd/glue.d:1187]: I was talking about this branching, perhaps use `static if` to avoid uncovered lines of unreachable code.

**Comment by maxhaton** [comment]
This will be tested once it's actually enabled for the Monterey target
(i.e. tests for Monterey will pass at all).

As for other Itanium targets i guess we can permute arguments if needed.

File: src/dmd/glue.d (hunk 4 above)
Comment (ibuclaw) [reply src/dmd/glue.d:1187]: Code is visibly not unreachable though, and if it gets flagged then that's just a limitation in the coverage report.

(That this bool is always true just tells me that this PR is still WIP and should really be a draft)

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Action items to tick off:
- [ ] Buffer on stack, not with calloc.
- [x] Put generated function in "C" static section (SCstatic?).
- [x] Give generated function an uncallable name (ie `CDFN.%u`).
- [x] Define new TargetC hook (`target.c.dtorsUseAtexit`), set it true for OSX, false for everyone else.
- [x] Add Monterey image to the CirrusCI pipeline.
- [ ] (Optionally) collect all dtor functions and wrap them into just one generated dtor thunk call that in turn calls all dtors in the order we define (do we trust atexit to really get this right?).
- [x] (Optionally) declare `extern(C) extern int __dso_handle;` and pass its address as the third parameter to cxa_atexit.

**Comment by maxhaton** [comment]
The logic is sound I'm just away for the weekend so I haven't pushed the
target specific logic yet. Draft or not doesn't mean ~~FA~~ anything, I
just wanted to test that it actually passes the tests before I went away.

File: .cirrus.yml [hunk 5]
```diff
@@ -82,6 +82,22 @@ task:
   << : *COMMON_STEPS_TEMPLATE
 
 # Mac
+task:
```
Comment (ljmf00) [inline .cirrus.yml:85]: We should change this to conform to the YAML standard. Cirrus CI is permissive about wrong YAML syntax, but duplicate mapping keys are not acceptable in pure YAML.

See https://github.com/dlang/phobos/commit/05c1b59919d1f2f55023ecb38ba1959216ae4235#diff-62587956f943bb2503db7bc6dd27d0d888074a1c0ecaab3f570ad611aff0f7bb and https://twitter.com/codewithanthony/status/1430393351212355585 .

**Comment by maxhaton** [comment]
@ibuclaw 
```
Test 'runnable/test17868.d' failed. The logged output:
/private/var/folders/xx/vl5f934s6k927z1vyyl9cxth0000gn/T/cirrus-ci-build/generated/osx/release/64/dmd -conf= -m64 -Irunnable -betterC  -fPIC  -odtest_results/runnable -oftest_results/runnable/test17868_0  runnable/test17868.d 
Undefined symbols for architecture x86_64:
  "__dso_handle", referenced from:
      __dmd_crt_destructor_thunk.0 in test17868_0.o
      __dmd_crt_destructor_thunk.1 in test17868_0.o
```

**Comment by ibuclaw** [comment]
> @ibuclaw 
> ```
> Test 'runnable/test17868.d' failed. The logged output:
> /private/var/folders/xx/vl5f934s6k927z1vyyl9cxth0000gn/T/cirrus-ci-build/generated/osx/release/64/dmd -conf= -m64 -Irunnable -betterC  -fPIC  -odtest_results/runnable -oftest_results/runnable/test17868_0  runnable/test17868.d 
> Undefined symbols for architecture x86_64:
>   "__dso_handle", referenced from:
>       __dmd_crt_destructor_thunk.0 in test17868_0.o
>       __dmd_crt_destructor_thunk.1 in test17868_0.o
> ```

There's probably something your doing wrong with name generation? OSX prefixes symbols with an extra `_`.

File: src/dmd/glue.d [hunk 6]
```diff
@@ -1163,8 +1163,72 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
+            char* buf = cast(char*) calloc(50, 1);
+            assert(buf);
+            const ret = snprintf(buf, 100, "_dmd_crt_destructor_thunk.%u", nthDestructor++);
+            assert(ret >= 0 && ret < 100, "snprintf either failed or overran buffer");
+            //Function symbol
+            auto newConstructor = symbol_calloc(buf);
+            //Build type
+            newConstructor.Stype = type_function(TYnfunc, [], false, type_alloc(TYvoid));
+            //Tell it it's supposed to be a C function. Does it do anything? Not sure.
+            newConstructor.Stype.Tmangle = mTYman_c;
```
Comment (ibuclaw) [inline src/dmd/glue.d:1198]: I think you probably need to set this for `__dso_handle` too.

Comment (maxhaton) [reply src/dmd/glue.d:1198]: I'm not convinced this actually does anything. If I let it generate the symbol itself on Linux it links just fine. I was reading an LLD thread and I think the symbol is supposed to be generated by the linker on MacOS

Comment (ibuclaw) [reply src/dmd/glue.d:1198]: > I'm not convinced this actually does anything. If I let it generate the symbol itself on Linux it links just fine. I was reading an LLD thread and I think the symbol is supposed to be generated by the linker on MacOS

Yes it does, for instance ensures that the proper mangling is done on the symbol.

**Comment by maxhaton** [comment]
@ibuclaw seems to be working

File: src/dmd/glue.d [hunk 7]
```diff
@@ -1163,8 +1163,74 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
...
+            func_t* funcState = newConstructor.Sfunc;
+            //Init start block
+            funcState.Fstartblock = block_calloc();
+            block* startBlk = funcState.Fstartblock;
+            //Make that block run __cxa_atexit(&func);
+            auto atexitSym = getRtlsym(RTLSYM.CXA_ATEXIT);
+            Symbol* dso_handle = symbol_calloc("__dso_handle");
+            dso_handle.Stype = type_fake(TYint);
+            //Try to get MacOS _ prefix-ism right.
+            type_setmangle(&dso_handle.Stype, mTYman_c);
```
Comment (ibuclaw) [inline src/dmd/glue.d:1211]: You use `type_setmangle` here, but set `newConstructor.Stype.Tmangle` directly above, is it really done different between functions and variables?

File: src/dmd/glue.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/glue.d:1174]: @maxhaton are you going to fix this comment?

**Comment by maxhaton** [comment]
That slipped my mind, let me fix it shortly.

**Review by ibuclaw — CHANGES_REQUESTED** [review]
One nit left as far as I can tell.

The high level looks OK to me, this matches gcc and clang behaviour on Darwin. I'm in no position to verify that the implementation is correct for dmd though (@WalterBright do you want to give this a brief look over?)

File: src/dmd/target.d [hunk 8]
```diff
@@ -1245,6 +1245,13 @@ struct TargetC
             bitFieldStyle = BitFieldStyle.Gcc_Clang;
         else
             assert(0);
+        /*
+            MacOS Monterey (12) does not support C runtime destructors.
```
Comment (ibuclaw) [inline src/dmd/target.d:1249]: Strictly, it's not that they don't support, they choose to do so using atexit instead of having a dedicated binary section.

**Review by ibuclaw — APPROVED** [review]
Applying 72h label, hopefully someone will respond in that time.

**Comment by ljmf00** [comment]
> Applying 72h label, hopefully someone will respond in that time.

The freshly added CI is failing.

**Comment by maxhaton** [comment]
> > Applying 72h label, hopefully someone will respond in that time.
> 
> The freshly added CI is failing.

Rerunning because unless it's something else it was one of those stupid Phobos failures

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**